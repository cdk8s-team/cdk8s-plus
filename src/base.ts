import { ApiObjectMetadata, ApiObject, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { IApiResource, IApiEndpoint } from './api-resource.generated';
import * as r from './role';
import * as rb from './role-binding';

/**
 * Initialization properties for resources.
 */
export interface ResourceProps {
  /**
   * Metadata that all persisted resources must have, which includes all objects
   * users must create.
   */
  readonly metadata?: ApiObjectMetadata;
}

/**
 * Represents a resource.
 */
export interface IResource {
  /**
   * The Kubernetes name of this resource.
   */
  readonly name: string;

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  readonly apiVersion: string;

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io")
   */
  readonly apiGroup: string;

  /**
   * The object kind (e.g. "Deployment").
   */
  readonly kind: string;
}

/**
 * Base class for all Kubernetes objects in stdk8s. Represents a single
 * resource.
 */
export abstract class Resource extends Construct implements IResource, IApiResource, IApiEndpoint {

  /**
   * The underlying cdk8s API object.
   */
  protected abstract readonly apiObject: ApiObject;

  public get metadata(): ApiObjectMetadataDefinition {
    return this.apiObject.metadata;
  }

  /**
   * The name of this API object.
   */
  public get name(): string {
    return this.apiObject.name;
  }

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  public get apiVersion(): string {
    return this.apiObject.apiVersion;
  }

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io").
   */
  public get apiGroup(): string {
    return this.apiObject.apiGroup;
  }

  /**
   * The object kind (e.g. "Deployment").
   */
  public get kind(): string {
    return this.apiObject.kind;
  }

  public get resourceName(): string | undefined {
    return this.name;
  }

  public readonly abstract resourceType: string;

  public asApiResource(): IApiResource | undefined {
    return this;
  }

  public asNonApiResource(): string | undefined {
    return undefined;
  }

  /**
   * Grants the list of subjects permissions to read this resource.
   */
  public grantRead(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowRead(this);
    return role.bind(...subjects);
  }

  /**
   * Grants the list of subjects permissions to read and write this resource.
   */
  public grantReadWrite(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowReadWrite(this);
    return role.bind(...subjects);
  }

}