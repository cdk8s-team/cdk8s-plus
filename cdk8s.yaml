output: dist
pluginsDirectory: /home/runner/.cdk8s/plugins
imports:
  - k8s@1.27.0
  - k8s@1.28.0
