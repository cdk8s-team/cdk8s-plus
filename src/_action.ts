import { Container } from './container';
import * as k8s from './imports/k8s';

/**
 * Utility class to implement the conversion between our API and the k8s action
 * structure. Used both for probes and handlers.
 *
 * @internal
 */
export class Action {

  public static fromTcpSocket(container: Container, options: { port?: number; host?: string } = {}): k8s.TcpSocketAction {
    return {
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
      host: options.host,
    };
  }

  public static fromCommand(command: string[]): k8s.ExecAction {
    return { command };
  }

  public static fromHttpGet(container: Container, path: string, options: { port?: number } = { }): k8s.HttpGetAction {
    return {
      path,
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
    };
  }
}