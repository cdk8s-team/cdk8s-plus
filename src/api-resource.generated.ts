// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

/**
 * Represents a resource or collection of resources.
 */
export interface IApiResource {
  /**
   * The group portion of the API version (e.g. `authorization.k8s.io`).
   */
  readonly apiGroup: string;

  /**
   * The name of a resource type as it appears in the relevant API endpoint.
   * @example - "pods" or "pods/log"
   * @see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources
   */
  readonly resourceType: string;

  /**
   * The unique, namespace-global, name of an object inside the Kubernetes cluster.
   *
   * If this is omitted, the ApiResource should represent all objects of the given type.
   */
  readonly resourceName?: string;
}
/**
 * An API Endpoint can either be a resource descriptor (e.g /pods)
 * or a non resource url (e.g /healthz). It must be one or the other, and not both.
 */
export interface IApiEndpoint {

  /**
   * Return the IApiResource this object represents.
   */
  asApiResource(): IApiResource | undefined;

  /**
   * Return the non resource url this object represents.
   */
  asNonApiResource(): string | undefined;

}

/**
 * Options for `ApiResource`.
 */
export interface ApiResourceOptions {
  /**
   * The group portion of the API version (e.g. `authorization.k8s.io`).
   */
  readonly apiGroup: string;

  /**
   * The name of the resource type as it appears in the relevant API endpoint.
   * @example - "pods" or "pods/log"
   * @see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources
   */
  readonly resourceType: string;
}

/**
 * Represents information about an API resource type.
 */
export class ApiResource implements IApiResource, IApiEndpoint {
  /**
   * API resource information for Binding.
   */
  public static readonly BINDINGS = new ApiResource({
    apiGroup: '',
    resourceType: 'bindings',
  });

  /**
   * API resource information for ComponentStatus.
   */
  public static readonly COMPONENT_STATUSES = new ApiResource({
    apiGroup: '',
    resourceType: 'componentstatuses',
  });

  /**
   * API resource information for ConfigMap.
   */
  public static readonly CONFIG_MAPS = new ApiResource({
    apiGroup: '',
    resourceType: 'configmaps',
  });

  /**
   * API resource information for Endpoints.
   */
  public static readonly ENDPOINTS = new ApiResource({
    apiGroup: '',
    resourceType: 'endpoints',
  });

  /**
   * API resource information for Event.
   */
  public static readonly EVENTS = new ApiResource({
    apiGroup: '',
    resourceType: 'events',
  });

  /**
   * API resource information for LimitRange.
   */
  public static readonly LIMIT_RANGES = new ApiResource({
    apiGroup: '',
    resourceType: 'limitranges',
  });

  /**
   * API resource information for Namespace.
   */
  public static readonly NAMESPACES = new ApiResource({
    apiGroup: '',
    resourceType: 'namespaces',
  });

  /**
   * API resource information for Node.
   */
  public static readonly NODES = new ApiResource({
    apiGroup: '',
    resourceType: 'nodes',
  });

  /**
   * API resource information for PersistentVolumeClaim.
   */
  public static readonly PERSISTENT_VOLUME_CLAIMS = new ApiResource({
    apiGroup: '',
    resourceType: 'persistentvolumeclaims',
  });

  /**
   * API resource information for PersistentVolume.
   */
  public static readonly PERSISTENT_VOLUMES = new ApiResource({
    apiGroup: '',
    resourceType: 'persistentvolumes',
  });

  /**
   * API resource information for Pod.
   */
  public static readonly PODS = new ApiResource({
    apiGroup: '',
    resourceType: 'pods',
  });

  /**
   * API resource information for PodTemplate.
   */
  public static readonly POD_TEMPLATES = new ApiResource({
    apiGroup: '',
    resourceType: 'podtemplates',
  });

  /**
   * API resource information for ReplicationController.
   */
  public static readonly REPLICATION_CONTROLLERS = new ApiResource({
    apiGroup: '',
    resourceType: 'replicationcontrollers',
  });

  /**
   * API resource information for ResourceQuota.
   */
  public static readonly RESOURCE_QUOTAS = new ApiResource({
    apiGroup: '',
    resourceType: 'resourcequotas',
  });

  /**
   * API resource information for Secret.
   */
  public static readonly SECRETS = new ApiResource({
    apiGroup: '',
    resourceType: 'secrets',
  });

  /**
   * API resource information for ServiceAccount.
   */
  public static readonly SERVICE_ACCOUNTS = new ApiResource({
    apiGroup: '',
    resourceType: 'serviceaccounts',
  });

  /**
   * API resource information for Service.
   */
  public static readonly SERVICES = new ApiResource({
    apiGroup: '',
    resourceType: 'services',
  });

  /**
   * API resource information for MutatingWebhookConfiguration.
   */
  public static readonly MUTATING_WEBHOOK_CONFIGURATIONS = new ApiResource({
    apiGroup: 'admissionregistration.k8s.io',
    resourceType: 'mutatingwebhookconfigurations',
  });

  /**
   * API resource information for ValidatingWebhookConfiguration.
   */
  public static readonly VALIDATING_WEBHOOK_CONFIGURATIONS = new ApiResource({
    apiGroup: 'admissionregistration.k8s.io',
    resourceType: 'validatingwebhookconfigurations',
  });

  /**
   * API resource information for CustomResourceDefinition.
   */
  public static readonly CUSTOM_RESOURCE_DEFINITIONS = new ApiResource({
    apiGroup: 'apiextensions.k8s.io',
    resourceType: 'customresourcedefinitions',
  });

  /**
   * API resource information for APIService.
   */
  public static readonly API_SERVICES = new ApiResource({
    apiGroup: 'apiregistration.k8s.io',
    resourceType: 'apiservices',
  });

  /**
   * API resource information for ControllerRevision.
   */
  public static readonly CONTROLLER_REVISIONS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'controllerrevisions',
  });

  /**
   * API resource information for DaemonSet.
   */
  public static readonly DAEMON_SETS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'daemonsets',
  });

  /**
   * API resource information for Deployment.
   */
  public static readonly DEPLOYMENTS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'deployments',
  });

  /**
   * API resource information for ReplicaSet.
   */
  public static readonly REPLICA_SETS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'replicasets',
  });

  /**
   * API resource information for StatefulSet.
   */
  public static readonly STATEFUL_SETS = new ApiResource({
    apiGroup: 'apps',
    resourceType: 'statefulsets',
  });

  /**
   * API resource information for TokenReview.
   */
  public static readonly TOKEN_REVIEWS = new ApiResource({
    apiGroup: 'authentication.k8s.io',
    resourceType: 'tokenreviews',
  });

  /**
   * API resource information for LocalSubjectAccessReview.
   */
  public static readonly LOCAL_SUBJECT_ACCESS_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'localsubjectaccessreviews',
  });

  /**
   * API resource information for SelfSubjectAccessReview.
   */
  public static readonly SELF_SUBJECT_ACCESS_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'selfsubjectaccessreviews',
  });

  /**
   * API resource information for SelfSubjectRulesReview.
   */
  public static readonly SELF_SUBJECT_RULES_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'selfsubjectrulesreviews',
  });

  /**
   * API resource information for SubjectAccessReview.
   */
  public static readonly SUBJECT_ACCESS_REVIEWS = new ApiResource({
    apiGroup: 'authorization.k8s.io',
    resourceType: 'subjectaccessreviews',
  });

  /**
   * API resource information for HorizontalPodAutoscaler.
   */
  public static readonly HORIZONTAL_POD_AUTOSCALERS = new ApiResource({
    apiGroup: 'autoscaling',
    resourceType: 'horizontalpodautoscalers',
  });

  /**
   * API resource information for CronJob.
   */
  public static readonly CRON_JOBS = new ApiResource({
    apiGroup: 'batch',
    resourceType: 'cronjobs',
  });

  /**
   * API resource information for Job.
   */
  public static readonly JOBS = new ApiResource({
    apiGroup: 'batch',
    resourceType: 'jobs',
  });

  /**
   * API resource information for CertificateSigningRequest.
   */
  public static readonly CERTIFICATE_SIGNING_REQUESTS = new ApiResource({
    apiGroup: 'certificates.k8s.io',
    resourceType: 'certificatesigningrequests',
  });

  /**
   * API resource information for Lease.
   */
  public static readonly LEASES = new ApiResource({
    apiGroup: 'coordination.k8s.io',
    resourceType: 'leases',
  });

  /**
   * API resource information for EndpointSlice.
   */
  public static readonly ENDPOINT_SLICES = new ApiResource({
    apiGroup: 'discovery.k8s.io',
    resourceType: 'endpointslices',
  });

  /**
   * API resource information for FlowSchema.
   */
  public static readonly FLOW_SCHEMAS = new ApiResource({
    apiGroup: 'flowcontrol.apiserver.k8s.io',
    resourceType: 'flowschemas',
  });

  /**
   * API resource information for PriorityLevelConfiguration.
   */
  public static readonly PRIORITY_LEVEL_CONFIGURATIONS = new ApiResource({
    apiGroup: 'flowcontrol.apiserver.k8s.io',
    resourceType: 'prioritylevelconfigurations',
  });

  /**
   * API resource information for IngressClass.
   */
  public static readonly INGRESS_CLASSES = new ApiResource({
    apiGroup: 'networking.k8s.io',
    resourceType: 'ingressclasses',
  });

  /**
   * API resource information for Ingress.
   */
  public static readonly INGRESSES = new ApiResource({
    apiGroup: 'networking.k8s.io',
    resourceType: 'ingresses',
  });

  /**
   * API resource information for NetworkPolicy.
   */
  public static readonly NETWORK_POLICIES = new ApiResource({
    apiGroup: 'networking.k8s.io',
    resourceType: 'networkpolicies',
  });

  /**
   * API resource information for RuntimeClass.
   */
  public static readonly RUNTIME_CLASSES = new ApiResource({
    apiGroup: 'node.k8s.io',
    resourceType: 'runtimeclasses',
  });

  /**
   * API resource information for PodDisruptionBudget.
   */
  public static readonly POD_DISRUPTION_BUDGETS = new ApiResource({
    apiGroup: 'policy',
    resourceType: 'poddisruptionbudgets',
  });

  /**
   * API resource information for PodSecurityPolicy.
   */
  public static readonly POD_SECURITY_POLICIES = new ApiResource({
    apiGroup: 'policy',
    resourceType: 'podsecuritypolicies',
  });

  /**
   * API resource information for ClusterRoleBinding.
   */
  public static readonly CLUSTER_ROLE_BINDINGS = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'clusterrolebindings',
  });

  /**
   * API resource information for ClusterRole.
   */
  public static readonly CLUSTER_ROLES = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'clusterroles',
  });

  /**
   * API resource information for RoleBinding.
   */
  public static readonly ROLE_BINDINGS = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'rolebindings',
  });

  /**
   * API resource information for Role.
   */
  public static readonly ROLES = new ApiResource({
    apiGroup: 'rbac.authorization.k8s.io',
    resourceType: 'roles',
  });

  /**
   * API resource information for PriorityClass.
   */
  public static readonly PRIORITY_CLASSES = new ApiResource({
    apiGroup: 'scheduling.k8s.io',
    resourceType: 'priorityclasses',
  });

  /**
   * API resource information for CSIDriver.
   */
  public static readonly CSI_DRIVERS = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'csidrivers',
  });

  /**
   * API resource information for CSINode.
   */
  public static readonly CSI_NODES = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'csinodes',
  });

  /**
   * API resource information for CSIStorageCapacity.
   */
  public static readonly CSI_STORAGE_CAPACITIES = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'csistoragecapacities',
  });

  /**
   * API resource information for StorageClass.
   */
  public static readonly STORAGE_CLASSES = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'storageclasses',
  });

  /**
   * API resource information for VolumeAttachment.
   */
  public static readonly VOLUME_ATTACHMENTS = new ApiResource({
    apiGroup: 'storage.k8s.io',
    resourceType: 'volumeattachments',
  });

  /**
   * API resource information for a custom resource type.
   */
  public static custom(options: ApiResourceOptions): ApiResource {
    return new ApiResource(options);
  };

  /**
   * The group portion of the API version (e.g. `authorization.k8s.io`).
   */
  public readonly apiGroup: string;

  /**
   * The name of the resource type as it appears in the relevant API endpoint.
   * @example - "pods" or "pods/log"
   * @see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources
   */
  public readonly resourceType: string;

  public asApiResource(): IApiResource | undefined {
    return this;
  }

  public asNonApiResource(): string | undefined {
    return undefined;
  }
  private constructor(options: ApiResourceOptions) {
    this.apiGroup = options.apiGroup;
    this.resourceType = options.resourceType;
  }
}

/**
 * Factory for creating non api resources.
 */
export class NonApiResource implements IApiEndpoint {

  public static of(url: string): NonApiResource {
    return new NonApiResource(url);
  }

  private constructor(private readonly nonResourceUrl: string) {};
  public asApiResource(): IApiResource | undefined {
    return undefined;
  }

  public asNonApiResource(): string | undefined {

    return this.nonResourceUrl;
  }
}