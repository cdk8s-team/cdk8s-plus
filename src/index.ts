export * from './base';
export * from './config-map';
export * from './container';
export * from './deployment';
export * from './job';
export * from './pod';
export * from './secret';
export * from './service-account';
export * from './service';
export * from './stateful-set';
export * from './volume';
export * from './ingress';
export * from './probe';
export * from './pvc';
export * from './pv';
export * from './handler';
export * from './workload';
export * from './daemon-set';
export * from './role';
export * from './role-binding';
export * from './namespace';

export * from './api-resource.generated';
