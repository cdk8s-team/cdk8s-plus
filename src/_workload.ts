import { ApiObjectMetadata, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import { Container, ContainerProps } from './container';
import { HostAlias, PodSecurityContext, PodSpec, PodSpecProps, RestartPolicy } from './pod';
import { IServiceAccount } from './service-account';
import { Volume } from './volume';

/**
 * Properties for `Workload`.
 */
export interface WorkloadProps extends ResourceProps, PodSpecProps {

  /**
   * The pod metadata of this workload.
   */
  readonly podMetadata?: ApiObjectMetadata;

}

export abstract class Workload extends Resource {

  protected readonly podSpec: PodSpec;

  /**
   * The metadata of pods in this deployment.
   */
  public readonly podMetadata: ApiObjectMetadataDefinition;

  constructor(scope: Construct, id: string, props: WorkloadProps = {}) {
    super(scope, id);

    this.podSpec = new PodSpec(props);
    this.podMetadata = new ApiObjectMetadataDefinition(props.podMetadata);
  }

  public get containers(): Container[] {
    return this.podSpec.containers;
  }

  public get initContainers(): Container[] {
    return this.podSpec.initContainers;
  }

  public get volumes(): Volume[] {
    return this.podSpec.volumes;
  }

  public get restartPolicy(): RestartPolicy | undefined {
    return this.podSpec.restartPolicy;
  }

  public get serviceAccount(): IServiceAccount | undefined {
    return this.podSpec.serviceAccount;
  }

  public get securityContext(): PodSecurityContext {
    return this.podSpec.securityContext;
  }

  public get hostAliases(): HostAlias[] {
    return this.podSpec.hostAliases;
  }

  public addContainer(container: ContainerProps): Container {
    return this.podSpec.addContainer(container);
  }

  public addInitContainer(container: ContainerProps): Container {
    return this.podSpec.addInitContainer(container);
  }

  public addVolume(volume: Volume): void {
    return this.podSpec.addVolume(volume);
  }

  public addHostAlias(hostAlias: HostAlias): void {
    return this.podSpec.addHostAlias(hostAlias);
  }

}