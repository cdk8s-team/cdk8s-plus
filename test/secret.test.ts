import { Testing, ApiObject } from 'cdk8s';
import { Node } from 'constructs';
import * as kplus from '../src';

test('can grant permissions on imported', () => {

  const chart = Testing.chart();
  const secret = kplus.Secret.fromSecretName(chart, 'Secret', 'secret');

  const role = new kplus.Role(chart, 'Role');
  role.allowRead(secret);

  expect(Testing.synth(chart)).toMatchSnapshot();

});

test('defaultChild', () => {
  const chart = Testing.chart();

  const defaultChild = Node.of(new kplus.Secret(chart, 'Secret'))
    .defaultChild as ApiObject;

  expect(defaultChild.kind).toEqual('Secret');
});

test('Can be imported from secret name', () => {
  const chart = Testing.chart();
  const secret = kplus.Secret.fromSecretName(chart, 'Secret', 'secret');

  expect(secret.name).toEqual('secret');
});

test('Can create a new secret', () => {
  const chart = Testing.chart();

  new kplus.Secret(chart, 'Secret', {
    stringData: {
      key: 'value',
    },
    type: 'kubernetes.io/tls',
  });

  expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "test-secret-c837fa76",
    },
    "stringData": Object {
      "key": "value",
    },
    "type": "kubernetes.io/tls",
  },
]
`);
});

test('Can add data to new secrets', () => {
  const chart = Testing.chart();

  const secret = new kplus.Secret(chart, 'Secret');
  secret.addStringData('key', 'value');

  expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "test-secret-c837fa76",
    },
    "stringData": Object {
      "key": "value",
    },
  },
]
`);
});

test('Can create a basic auth secret', () => {
  const chart = Testing.chart();

  new kplus.BasicAuthSecret(chart, 'BasicAuthSecret', {
    username: 'admin',
    password: 't0p-Secret',
  });

  expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "test-basicauthsecret-c82606a8",
    },
    "stringData": Object {
      "password": "t0p-Secret",
      "username": "admin",
    },
    "type": "kubernetes.io/basic-auth",
  },
]
`);
});

test('can override the name of a basic auth secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.BasicAuthSecret(chart, 'BasicAuthSecret', {
    metadata: {
      name: 'override-name',
    },
    username: 'admin',
    password: 't0p-Secret',
  });

  const manifest = Testing.synth(chart);

  expect(secret.name).toEqual('override-name');
  expect(manifest[0].metadata.name).toEqual('override-name');

});

test('Can create an ssh auth secret', () => {
  const chart = Testing.chart();

  new kplus.SshAuthSecret(chart, 'SshAuthSecret', {
    sshPrivateKey: 'fake-private-key',
  });

  expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "test-sshauthsecret-c8356ec6",
    },
    "stringData": Object {
      "ssh-privatekey": "fake-private-key",
    },
    "type": "kubernetes.io/ssh-auth",
  },
]
`);
});

test('can override the name of an ssh auth secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.SshAuthSecret(chart, 'SshAuthSecret', {
    metadata: {
      name: 'override-name',
    },
    sshPrivateKey: 'fake-private-key',
  });

  const manifest = Testing.synth(chart);

  expect(secret.name).toEqual('override-name');
  expect(manifest[0].metadata.name).toEqual('override-name');

});

test('Can create a service account token secret', () => {
  const chart = Testing.chart();

  const sa = new kplus.ServiceAccount(chart, 'ServiceAccount');
  const secret = new kplus.ServiceAccountTokenSecret(chart, 'ServiceAccountToken', {
    serviceAccount: sa,
  });
  secret.addStringData('extra', 'foo');

  // the "token" key in "data" should be automatically filled in by a
  // Kubernetes controller
  expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "v1",
    "automountServiceAccountToken": false,
    "kind": "ServiceAccount",
    "metadata": Object {
      "name": "test-serviceaccount-c8f15383",
    },
  },
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "annotations": Object {
        "kubernetes.io/service-account.name": "test-serviceaccount-c8f15383",
      },
      "name": "test-serviceaccounttoken-c8eca6af",
    },
    "stringData": Object {
      "extra": "foo",
    },
    "type": "kubernetes.io/service-account-token",
  },
]
`);
});

test('can override the name of a service account token secret', () => {

  const chart = Testing.chart();

  const sa = new kplus.ServiceAccount(chart, 'ServiceAccount');
  const secret = new kplus.ServiceAccountTokenSecret(chart, 'ServiceAccountToken', {
    metadata: {
      name: 'override-name',
    },
    serviceAccount: sa,
  });
  secret.addStringData('extra', 'foo');

  const manifest = Testing.synth(chart);

  expect(secret.name).toEqual('override-name');
  expect(manifest[1].metadata.name).toEqual('override-name');

});

test('can add annotations to a service account token secret', () => {

  const chart = Testing.chart();

  const sa = new kplus.ServiceAccount(chart, 'ServiceAccount');
  const secret = new kplus.ServiceAccountTokenSecret(chart, 'ServiceAccountToken', {
    metadata: {
      annotations: {
        'cdk8s.io/hello': 'world',
      },
    },
    serviceAccount: sa,
  });
  secret.addStringData('extra', 'foo');

  const manifest = Testing.synth(chart);

  expect(manifest[1].metadata.annotations).toStrictEqual({
    'kubernetes.io/service-account.name': 'test-serviceaccount-c8f15383',
    'cdk8s.io/hello': 'world',
  });

});

test('Can create a TLS secret', () => {
  const chart = Testing.chart();

  new kplus.TlsSecret(chart, 'TlsSecret', {
    tlsCert: 'tls-cert-value',
    tlsKey: 'tls-key-value',
  });

  expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "test-tlssecret-c8c8af35",
    },
    "stringData": Object {
      "tls.crt": "tls-cert-value",
      "tls.key": "tls-key-value",
    },
    "type": "kubernetes.io/tls",
  },
]
`);
});

test('can override the name of a tls secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.TlsSecret(chart, 'TlsSecret', {
    metadata: {
      name: 'override-name',
    },
    tlsCert: 'tls-cert-value',
    tlsKey: 'tls-key-value',
  });

  const manifest = Testing.synth(chart);

  expect(secret.name).toEqual('override-name');
  expect(manifest[0].metadata.name).toEqual('override-name');

});

test('Can create a Docker config secret', () => {
  const chart = Testing.chart();

  new kplus.DockerConfigSecret(chart, 'DockerConfigSecret', {
    data: {
      auths: {
        'hub.xxx.com': {
          username: 'xxx',
          password: 'xxx',
          email: 'xxx',
          auth: 'xxx',
        },
      },
    },
  });

  expect(Testing.synth(chart)).toMatchInlineSnapshot(`
Array [
  Object {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": Object {
      "name": "test-dockerconfigsecret-c8b65039",
    },
    "stringData": Object {
      ".dockerconfigjson": "{\\"auths\\":{\\"hub.xxx.com\\":{\\"username\\":\\"xxx\\",\\"password\\":\\"xxx\\",\\"email\\":\\"xxx\\",\\"auth\\":\\"xxx\\"}}}",
    },
    "type": "kubernetes.io/dockerconfigjson",
  },
]
`);
});

test('can override the name of a DockerConfig secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.DockerConfigSecret(chart, 'DockerConfigSecret', {
    metadata: {
      name: 'override-name',
    },
    data: {
      auths: {
        'hub.xxx.com': {
          username: 'xxx',
          password: 'xxx',
          email: 'xxx',
          auth: 'xxx',
        },
      },
    },
  });

  const manifest = Testing.synth(chart);

  expect(secret.name).toEqual('override-name');
  expect(manifest[0].metadata.name).toEqual('override-name');

});

test('default immutability', () => {

  const chart = Testing.chart();

  const secret = new kplus.Secret(chart, 'Secret');

  const spec = Testing.synth(chart)[0];

  expect(secret.immutable).toBeFalsy();
  expect(spec.immutable).toBeFalsy();

});

test('can configure an immutable generic secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.Secret(chart, 'Secret', {
    immutable: true,
  });

  const spec = Testing.synth(chart)[0];

  expect(secret.immutable).toBeTruthy();
  expect(spec.immutable).toBeTruthy();

});

test('can configure an immutable basic auth secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.BasicAuthSecret(chart, 'Secret', {
    username: 'user',
    password: 'pass',
    immutable: true,
  });

  const spec = Testing.synth(chart)[0];

  expect(secret.immutable).toBeTruthy();
  expect(spec.immutable).toBeTruthy();

});

test('can configure an immutable ssh auth secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.SshAuthSecret(chart, 'Secret', {
    sshPrivateKey: 'private',
    immutable: true,
  });

  const spec = Testing.synth(chart)[0];

  expect(secret.immutable).toBeTruthy();
  expect(spec.immutable).toBeTruthy();

});

test('can configure an immutable service account token secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.ServiceAccountTokenSecret(chart, 'Secret', {
    serviceAccount: kplus.ServiceAccount.fromServiceAccountName(chart, 'SA', 'sa'),
    immutable: true,
  });

  const spec = Testing.synth(chart)[0];

  expect(secret.immutable).toBeTruthy();
  expect(spec.immutable).toBeTruthy();

});

test('can configure an immutable tls secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.TlsSecret(chart, 'Secret', {
    tlsCert: 'cert',
    tlsKey: 'key',
    immutable: true,
  });

  const spec = Testing.synth(chart)[0];

  expect(secret.immutable).toBeTruthy();
  expect(spec.immutable).toBeTruthy();

});

test('can configure an immutable docker config secret', () => {

  const chart = Testing.chart();

  const secret = new kplus.DockerConfigSecret(chart, 'Secret', {
    data: {},
    immutable: true,
  });

  const spec = Testing.synth(chart)[0];

  expect(secret.immutable).toBeTruthy();
  expect(spec.immutable).toBeTruthy();

});