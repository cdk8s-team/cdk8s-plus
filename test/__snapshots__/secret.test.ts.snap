// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can grant permissions on imported 1`] = `
Array [
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "Role",
    "metadata": Object {
      "name": "test-role-c8eba390",
    },
    "rules": Array [
      Object {
        "apiGroups": Array [
          "",
        ],
        "resourceNames": Array [
          "secret",
        ],
        "resources": Array [
          "secrets",
        ],
        "verbs": Array [
          "get",
          "list",
          "watch",
        ],
      },
    ],
  },
]
`;
